tools/topology/topology2/platform/intel/dmic-generic.conf

```diff
@@ -211,175 +143,428 @@ IncludeByKey.PASSTHROUGH {
+								}
+							]
+							Object.Base.output_audio_format [
+								{
+									out_rate		$DMIC0_RATE
+									out_channels		4
+									out_bit_depth		32
+									out_valid_bit_depth	32
+									out_ch_cfg	$CHANNEL_CONFIG_3_POINT_1
+									out_ch_map	$CHANNEL_MAP_3_POINT_1
+								}
+							]
 						}
-						{
-							out_rate		$DMIC0_RATE
-							out_channels		4
-							out_bit_depth		32
-							out_valid_bit_depth	32
-							out_ch_cfg	$CHANNEL_CONFIG_3_POINT_1
-							out_ch_map	$CHANNEL_MAP_3_POINT_1
+					}
+				}
+
+				Object.Widget.gain.1 {
+					curve_duration 500000
+					num_input_audio_formats 2
+					num_output_audio_formats 2
+					num_input_audio_formats	    1
+					num_output_audio_formats    1
+					IncludeByKey.NUM_DMICS {
+						"2" {
+							Object.Base.input_audio_format [
+								{
+									in_rate			$DMIC0_RATE
+									in_bit_depth		32
+									in_valid_bit_depth	32
+								}
+							]
+							Object.Base.output_audio_format [
+								{
+									out_rate		$DMIC0_RATE
+									out_bit_depth		32
+									out_valid_bit_depth	32
+								}
+							]
+						}
+						"4" {
+							Object.Base.input_audio_format [
+								{
+									in_rate			$DMIC0_RATE
+									in_channels		4
+									in_bit_depth		32
+									in_valid_bit_depth	32
+									in_ch_cfg	$CHANNEL_CONFIG_3_POINT_1
+									in_ch_map	$CHANNEL_MAP_3_POINT_1
+								}
+							]
+							Object.Base.output_audio_format [
+								{
+									out_rate		$DMIC0_RATE
+									out_channels		4
+									out_bit_depth		32
+									out_valid_bit_depth	32
+									out_ch_cfg	$CHANNEL_CONFIG_3_POINT_1
+									out_ch_map	$CHANNEL_MAP_3_POINT_1
+								}
+							]
 						}
-					]
+					}
+					Object.Control.mixer.1 {
+						name 'Dmic0 Capture Volume'
+					}
+					Object.Control.mixer.2 {
+						name 'Dmic0 Capture Switch'
+
+						mute_led_use 		1
+						mute_led_direction	1
+					}
 				}
+
+
 				Object.Widget.pipeline."1" {
 					core $DMIC_CORE_ID
 					priority $DMIC_PIPELINE_PRIORITY
 				}
 			}
 		]
-		Object.Base.route [
-			{
-				source $DMIC0_DAI_COPIER
-				sink $DMIC0_DAI_GAIN
+
+		IncludeByKey.DMIC0_ENHANCED_CAPTURE {
+			"true"	{
+
+                                Object.Widget.tdfb.1 {
+					index $DMIC0_HOST_PIPELINE_ID
+					Object.Control {
+						bytes."1" {
+							name 'Dmic0 Capture TDFB bytes'
+							max 16384
+							IncludeByKey.EFX_DMIC0_TDFB_PARAMS {
+								"line2_pass" "include/components/tdfb/line2_pass.conf"
+								"line2_generic_pm10deg" "include/components/tdfb/line2_generic_pm10deg_48khz.conf"
+								"line2_50mm" "include/components/tdfb/line2_50mm_pm0_30_90deg_48khz.conf"
+								"line2_68mm" "include/components/tdfb/line2_68mm_pm0_30_90deg_48khz.conf"
+								"line2_74mm" "include/components/tdfb/line2_74mm_pm0_30_90deg_48khz.conf"
+								"line4_pass" "include/components/tdfb/line4_pass.conf"
+								"line4to2_pass" "include/components/tdfb/line4to2_pass.conf"
```

Comment:
Note:  For now only pass-through for four DMIC systems for 4ch capture. I don't have yet a 3 beams blob for TDFB but it should be possible. With such we could produce a 3.1  (L, R, C, LFE) or 3.0 format stream in 4ch container. Create of LFE stream is not possible with current TDFB so it would be zero PCM code. I'm not sure if 3.0 format wav is 4ch or 3ch. If 3ch I'm not sure it's possible with SOF. Also I'm not sure how Linux user space handles 4ch streams, does the SOF topology channel map type impact.

Reply:
Note: The blobs line4to2_pass and line4_generic_pm10deg convert 4ch to 2ch. To use them the topology needs to be set for 2ch PCM for DMIC.